library/psa_crypto.c

```diff
@@ -5474,6 +5474,137 @@ static psa_status_t psa_key_derivation_tls12_ecjpake_to_pms_read(
 }
 #endif
 
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_HMAC)
+static psa_status_t psa_key_derivation_pbkdf2_generate_block(
+    psa_pbkdf2_key_derivation_t *pbkdf2,
+    psa_algorithm_t prf_alg,
+    uint8_t prf_output_length,
+    psa_key_attributes_t *attributes)
+{
+    psa_status_t status;
+    psa_mac_operation_t mac_operation = PSA_MAC_OPERATION_INIT;
+    size_t mac_output_length;
+    uint8_t U_i[PSA_MAC_MAX_SIZE];
+    uint8_t *U_accumulator = pbkdf2->output_block;
+    uint64_t i;
+    uint8_t block_counter[4];
+
+    mac_operation.is_sign = 1;
+    mac_operation.mac_size = prf_output_length;
+    MBEDTLS_PUT_UINT32_BE(pbkdf2->block_number, block_counter, 0);
+
+    status = psa_driver_wrapper_mac_sign_setup(&mac_operation,
+                                               attributes,
+                                               pbkdf2->password,
+                                               pbkdf2->password_length,
+                                               prf_alg);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_update(&mac_operation, pbkdf2->salt, pbkdf2->salt_length);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_update(&mac_operation, block_counter, sizeof(block_counter));
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_sign_finish(&mac_operation, U_i, sizeof(U_i),
+                                 &mac_output_length);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+
+    if (mac_output_length != prf_output_length) {
+        status = PSA_ERROR_CORRUPTION_DETECTED;
+        goto cleanup;
+    }
+
+    memcpy(U_accumulator, U_i, prf_output_length);
+
+    for (i = 1; i < pbkdf2->input_cost; i++) {
+        status = psa_driver_wrapper_mac_compute(attributes,
+                                                pbkdf2->password,
+                                                pbkdf2->password_length,
+                                                prf_alg, U_i, prf_output_length,
+                                                U_i, sizeof(U_i),
```

Comment:
@gilles-peskine-arm In the mbedtls implementation of `mac_compute`, `mac_output_length` is set to `mac_size`. Should I change the `mac_size` variable from `sizeof(U_i)` to `prf_output_length` in a subsequent PR?

Reply:
Here the `sizeof(U_i)` argument is the size of the buffer, it's not based on the size of the output. Of course there'll be a runtime error if the output is longer. It really needs to be the size of the buffer, not the expected size of the output: if you make a mistake and the buffer is not large enough, and you pass the expected size of the output, then nothing will catch the mistake. Whereas if you pass the buffer size, the mac computation function will check and will see that the buffer isn't large enough.

Reply:
> Here the `sizeof(U_i)` argument is the size of the buffer, it's not based on the size of the output. 

Yes this is the expected value. But in `psa_mac_compute_internal` also when calling the driver_wrapper function the `mac_size` is passed as `PSA_MAC_LENGTH()`. If we check the value of `mac_output_length` inside the loop then it is `sizeof(U_i)`.

Reply:
This is only for the default mbedtls implementation.

Reply:
Looking at the code of `psa_mac_compute_internal`, it does the size check:
```
    if (mac_size < operation_mac_size) {
        status = PSA_ERROR_BUFFER_TOO_SMALL;
        goto exit;
    }
```
After that point, it passes `operation_mac_size` to the lower-level functions, which is both guaranteed to be less than the buffer size (checked above) and guaranteed to be at least the size of the result (it's exactly the size of the result, that's how `operation_mac_size` is calculated).

>  If we check the value of mac_output_length inside the loop then it is sizeof(U_i)

Ah, I see. `mbedtls_psa_mac_compute` assumes that the caller has set the size argument to be exactly the same as the output. So it behaves correctly when called by our implementation, but not as a standalone driver. It's not really surprising because that code is the result of splitting our original monolithic implementation, and we don't test our implementation of the driver interface as such, we only test it via the library interface. 

Since we're relying on calling the driver interface, and we're likely to do this in other places (generally, we want drivers to be able to call other “lower-level” primitives via the driver interface, without going through key ids), we should fix this. But more ambitiously, that means we need to _test_ this. And that's [a lot of new work](https://github.com/Mbed-TLS/mbedtls/issues/7801)!

Reply:
Should I resolve this in my case as `mac_size = prf_output_length` for the PBKDF for now, and later when the other changes are in place then rework this to the correct value?

Reply:
For now, yes, a workaround in the PBKDF2 layer seems fine, as long as there's a comment to explain.

Reply:
I will add this in the CMAC PR.